concurrency: refactor LockStateInfo to display shared locks correctly with lock strength

lyang24 opened the pull request:
Fixes: https://github.com/cockroachdb/cockroach/issues/107766
This pr decouples the LockStateInfo to keyLocks struct as many to one relationship to better reflect shared locks. crdb_internal.cluster_locks
will be reflecting the correct lock holders and lock strengthes.

Release note: None.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/108543)
<!-- Reviewable:end -->

lyang24 commented:
a lot of changes in lock table need to rethink how to apply the logics, converting this pr to draft for now.

nvanbenschoten commented:
Hi @lyang24, many of the changes in this area have settled down and we're hoping to stabilize shared locks soon. It would be great to address https://github.com/cockroachdb/cockroach/issues/107766 as part of this. Do you have interest in rebasing this PR on master and picking it back up again?

lyang24 commented:
> Hi @lyang24, many of the changes in this area have settled down and we're hoping to stabilize shared locks soon. It would be great to address #107766 as part of this. Do you have interest in rebasing this PR on master and picking it back up again?

Yep sounds good! I will to pick this up again right after roachfest.

Commit:
Fixes: #107766

This pr decouples the LockStateInfo to keyLocks struct as many to one relationship to better reflect shared locks. crdb_internal.cluster_locks
will be reflecting the correct lock holders and lock strengthes.

Release note: None.

nvanbenschoten commented on pkg/kv/kvserver/concurrency/lock_table.go:
```diff
@@ -4384,3 +4385,17 @@ func MarkSkipLockedReplayError(cause error) error {
+// the lock holders.
+func (kl *keyLocks) maxLockStrength() (lock.Strength) {
+	str := lock.Shared
+	for e := kl.holders.Front(); e != nil; e = e.Next() {
```
If a lock has multiple holders (e.g. shared locks), we'll want to represent that in the `LockStateInfo`. To do that, we'll need to reorganize `LockStateInfo` to contain multiple holders. We can add a new field to the message which is a slice of `lock.Holder`, where `lock.Holder` is a new proto that contains `enginepb.TxnMeta`, `lock.Durability`, `lock.Strength`, and `time.Duration`.

For backward compatibility, we can continue to populate the `LockHolder`, `Durability`, and `HoldDuration` fields with the first holder.

Making these changes will address the "This doesn't work with multiple lock holders" comment up above.

lyang24 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108543)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks and @nvanbenschoten)

___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 4393 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108543#gh-1348989110:-Ng7kO25B_q_uBe4X6eO:b-w266n9) ([raw file](https://github.com/cockroachdb/cockroach/blob/6075357072724099b1663befd8a507db5d1c4fbb/pkg/kv/kvserver/concurrency/lock_table.go#L4393)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

If a lock has multiple holders (e.g. shared locks), we'll want to represent that in the `LockStateInfo`. To do that, we'll need to reorganize `LockStateInfo` to contain multiple holders. We can add a new field to the message which is a slice of `lock.Holder`, where `lock.Holder` is a new proto that contains `enginepb.TxnMeta`, `lock.Durability`, `lock.Strength`, and `time.Duration`.

For backward compatibility, we can continue to populate the `LockHolder`, `Durability`, and `HoldDuration` fields with the first holder.

Making these changes will address the "This doesn't work with multiple lock holders" comment up above.
</blockquote></details>

Thanks for the feedback - for `HoldDuration` the old ways is to calculate the max duration among the lock holders should we keep it that way for backward compatibility?
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 1932 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108543#-Ng7k60bAVLJBku0vsLy:-Ng7k60bAVLJBku0vsLz:b-5310id) ([raw file](https://github.com/cockroachdb/cockroach/blob/106ab359eae259b34310867c210be744ba37fb8d/pkg/kv/kvserver/concurrency/lock_table.go#L1932)):*
> ```Go
> 	}
> 
> 	if !kl.isLocked() {
> ```

up for discussion -  if the kl is not locked what should we return here



<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
Thanks for the patch @lyang24 and apologies for not getting back to you on the review sooner. I've left some review comments in-line, but at a high level, Im wondering if we should adapt this approach to display all lock holders for a given key in `crdb_internal.cluster_locks` instead of just the one with the strongest lock strength. 

> Reviewed 2 of 8 files at r1, 3 of 7 files at r2, 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108543#-:-Ni6TN2s47pW-qD6y5jv:b-dk3ubz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lyang24 and @nvanbenschoten)

___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 1932 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108543#-Ng7k60bAVLJBku0vsLy:-Ni6TAZF9ZbcbXDM6g2W:b1ufhdo) ([raw file](https://github.com/cockroachdb/cockroach/blob/106ab359eae259b34310867c210be744ba37fb8d/pkg/kv/kvserver/concurrency/lock_table.go#L1932)):*
<details><summary><i>Previously, lyang24 (Lanqing Yang) wrote…</i></summary><blockquote>

up for discussion -  if the kl is not locked what should we return here
</blockquote></details>

Should we preserve the old behaviour, of a nil lock holder and unreplicated durability here? 
___
*[`pkg/roachpb/data.proto` line 615 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/108543#-Ni6QphL793T0894dsFw:-Ni6QphL793T0894dsFx:bh5uuns) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e0c5034915c6f3174eaa98e09a3e4654a3cf26c/pkg/roachpb/data.proto#L615)):*
> ```proto
>   repeated kv.kvserver.concurrency.lock.Waiter waiters = 6 [(gogoproto.nullable) = false];
>   // Lock holder metadata.
>   repeated LockHolderInfo lock_holders = 7;
> ```

Another approach could be to decouple the `LockSatateInfo` proto and `keyLocks` struct, and instead return multiple `LockStateInfos` for a single `keyLocks` struct if there are multiple lock holders. This approach has the benefit of addressing the linked issue without changing the code to construct `crdb_internal.cluster_locks`. 

With this approach, we can list out multiple locks on a key instead of just the highest strength one. We also won't need to make changes to `crdb_internal.go` to get this behaviour, which is nice. How do you feel about switching to this approach instead? 
 
___
*[`pkg/sql/crdb_internal.go` line 7912 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/108543#-Ni6SInlBKjOzeLHxLdB:-Ni6SInlBKjOzeLHxLdC:b-dzr8jn) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e0c5034915c6f3174eaa98e09a3e4654a3cf26c/pkg/sql/crdb_internal.go#L7912)):*
> ```Go
> 			}
> 			var lockStrength lock.Strength
> 			for _, holderInfo := range curLock.LockHolders {
> ```

An extension to my comment above, I think we should list out all locks on a key, not just the strongest one. 
___
*[`pkg/sql/logictest/testdata/logic_test/select_for_share` line 56 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108543#-Ng7ksIf1_2Kjd01HYA1:-Ni6SdQb6jg7y_2_Lo4z:b-fart5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d926f1e51e23d477bcf3027adc6beac11513e58a/pkg/sql/logictest/testdata/logic_test/select_for_share#L56)):*
<details><summary><i>Previously, lyang24 (Lanqing Yang) wrote…</i></summary><blockquote>

0 row will appear under the behavior of the current logic which is concerning.
I there at this point
testuser and root grab the shared lock
and testuser2 is trying to grab exclusive lock but getting blocked
</blockquote></details>

I don't quite follow your comment here. 


<!-- Sent from Reviewable.io -->

lyang24 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108543)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`pkg/roachpb/data.proto` line 615 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/108543#-Ni6QphL793T0894dsFw:-NiMuw1xAF7B-3eE2eOr:b-611h26) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e0c5034915c6f3174eaa98e09a3e4654a3cf26c/pkg/roachpb/data.proto#L615)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Another approach could be to decouple the `LockSatateInfo` proto and `keyLocks` struct, and instead return multiple `LockStateInfos` for a single `keyLocks` struct if there are multiple lock holders. This approach has the benefit of addressing the linked issue without changing the code to construct `crdb_internal.cluster_locks`. 

With this approach, we can list out multiple locks on a key instead of just the highest strength one. We also won't need to make changes to `crdb_internal.go` to get this behaviour, which is nice. How do you feel about switching to this approach instead?
</blockquote></details>

I like this approach. changing crbd_internal.go is not easy. However will this duplicate lock waiters display in crdb internal? i.e. we have two lock holders with shared strength both of them will have the same waiters it will resulted in duplicated waiter row in crdb internal?
___
*[`pkg/sql/crdb_internal.go` line 7912 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/108543#-Ni6SInlBKjOzeLHxLdB:-NiMwdH4-iFMsspernHY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e0c5034915c6f3174eaa98e09a3e4654a3cf26c/pkg/sql/crdb_internal.go#L7912)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

An extension to my comment above, I think we should list out all locks on a key, not just the strongest one.
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/select_for_share` line 56 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108543#-Ng7ksIf1_2Kjd01HYA1:-NiMwfHp04xle3sn8AsM:b-a5nvf0) ([raw file](https://github.com/cockroachdb/cockroach/blob/d926f1e51e23d477bcf3027adc6beac11513e58a/pkg/sql/logictest/testdata/logic_test/select_for_share#L56)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I don't quite follow your comment here.
</blockquote></details>

oh nvm some how my local test had some glitches it when fine after i rebased to newer master


<!-- Sent from Reviewable.io -->

lyang24 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108543)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 1954 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/108543#-NiNNXZiBzBtyPwiS5U0:-NiNNXZiBzBtyPwiS5U1:bd3pllu) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2d58b3e4d53eca6743b000282180deac33affea/pkg/kv/kvserver/concurrency/lock_table.go#L1954)):*
> ```Go
> 		}
> 		lockMode := tl.getLockMode()
> 		if idx > 0 {
> ```

This is a hack to prevent display duplicated lockwaiters on crdb_internal when there are multiple shared lock holder - semantically  each shared lock holder should have the same waiters set
___
*[`pkg/kv/kvserver/concurrency/testdata/lock_table/query` line 100 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/108543#-NiNNKe7B9r0jb35Nhbz:-NiNNKe7B9r0jb35Nhc-:b-89n82m) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2d58b3e4d53eca6743b000282180deac33affea/pkg/kv/kvserver/concurrency/testdata/lock_table/query#L100)):*
> ```
> query span=a,f max-bytes=10 uncontended
> ----
> num locks: 1, bytes returned: 53, resume reason: RESUME_BYTE_LIMIT, resume span: {e-f}
> ```

I believe the resumed span change is the result of decoupling the lockstateinfo from keylocks struct


<!-- Sent from Reviewable.io -->

lyang24 reviewed (commented):
just points out some areas to help with in-depth reviews

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108543#-:-NiNOMWcBga-0E56G5Tw:b-a1fkhc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)



<!-- Sent from Reviewable.io -->

lyang24 commented:
> Thanks for the patch @lyang24 and apologies for not getting back to you on the review sooner. I've left some review comments in-line, but at a high level, Im wondering if we should adapt this approach to display all lock holders for a given key in `crdb_internal.cluster_locks` instead of just the one with the strongest lock strength.
> 
> > Reviewed 2 of 8 files at r1, 3 of 7 files at r2, 3 of 3 files at r3, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108543#-:-Ni6TN2s47pW-qD6y5jv:b-dk3ubz)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lyang24 and @nvanbenschoten)
> 
> _[`pkg/kv/kvserver/concurrency/lock_table.go` line 1932 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108543#-Ng7k60bAVLJBku0vsLy:-Ni6TAZF9ZbcbXDM6g2W:b1ufhdo) ([raw file](https://github.com/cockroachdb/cockroach/blob/106ab359eae259b34310867c210be744ba37fb8d/pkg/kv/kvserver/concurrency/lock_table.go#L1932)):_
> 
> Previously, lyang24 (Lanqing Yang) wrote…
> Should we preserve the old behaviour, of a nil lock holder and unreplicated durability here?
> 
> _[`pkg/roachpb/data.proto` line 615 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/108543#-Ni6QphL793T0894dsFw:-Ni6QphL793T0894dsFx:bh5uuns) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e0c5034915c6f3174eaa98e09a3e4654a3cf26c/pkg/roachpb/data.proto#L615)):_
> 
> > ```proto
> >   repeated kv.kvserver.concurrency.lock.Waiter waiters = 6 [(gogoproto.nullable) = false];
> >   // Lock holder metadata.
> >   repeated LockHolderInfo lock_holders = 7;
> > ```
> 
> Another approach could be to decouple the `LockSatateInfo` proto and `keyLocks` struct, and instead return multiple `LockStateInfos` for a single `keyLocks` struct if there are multiple lock holders. This approach has the benefit of addressing the linked issue without changing the code to construct `crdb_internal.cluster_locks`.
> 
> With this approach, we can list out multiple locks on a key instead of just the highest strength one. We also won't need to make changes to `crdb_internal.go` to get this behaviour, which is nice. How do you feel about switching to this approach instead?
> 
> _[`pkg/sql/crdb_internal.go` line 7912 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/108543#-Ni6SInlBKjOzeLHxLdB:-Ni6SInlBKjOzeLHxLdC:b-dzr8jn) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e0c5034915c6f3174eaa98e09a3e4654a3cf26c/pkg/sql/crdb_internal.go#L7912)):_
> 
> > ```go
> > 			}
> > 			var lockStrength lock.Strength
> > 			for _, holderInfo := range curLock.LockHolders {
> > ```
> 
> An extension to my comment above, I think we should list out all locks on a key, not just the strongest one.
> 
> _[`pkg/sql/logictest/testdata/logic_test/select_for_share` line 56 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108543#-Ng7ksIf1_2Kjd01HYA1:-Ni6SdQb6jg7y_2_Lo4z:b-fart5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d926f1e51e23d477bcf3027adc6beac11513e58a/pkg/sql/logictest/testdata/logic_test/select_for_share#L56)):_
> 
> Previously, lyang24 (Lanqing Yang) wrote…
> I don't quite follow your comment here.

Hey @arulajmani, I have addressed the comments do you mind take another look? Also one of the concern of decoupling LockStateInfo to KeyLocks is the waiters of shared locks may become duplicated rows on crdb_internal.cluster_locks. I did do a hack to prevent it but now individual lockStateInfo of the same shared lock may not have the correct waiter information.

arulajmani reviewed (commented):
This is coming together quite nicely. I left a couple of comments inline, but this should be good once those are addressed.

> Reviewed 1 of 5 files at r4, 2 of 4 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108543#-:-NikckOA2kG3NPdeuVaA:b-ya31po)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lyang24 and @nvanbenschoten)

___
*[`-- commits` line 4 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/108543#-NikUa542fWprkor96BU):*
nit: s/decouple/decouples/g
___
*[`pkg/kv/kvserver/concurrency/lock_table.go` line 1954 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/108543#-NiNNXZiBzBtyPwiS5U0:-Nikc0fhBSGI5gJgIAGT:bmoajgu) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2d58b3e4d53eca6743b000282180deac33affea/pkg/kv/kvserver/concurrency/lock_table.go#L1954)):*
<details><summary><i>Previously, lyang24 (Lanqing Yang) wrote…</i></summary><blockquote>

This is a hack to prevent display duplicated lockwaiters on crdb_internal when there are multiple shared lock holder - semantically  each shared lock holder should have the same waiters set
</blockquote></details>

This works, but like you pointed out below, it means that we're returning a different set of waiters for the same key which isn't great. An alternative here would to return the correct waiter set for all locks and then do the filtering on the client instead.
___
*[`pkg/roachpb/data.proto` line 615 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/108543#-Ni6QphL793T0894dsFw:-Nikd8If6F3IMBQjLBZM:b13mk7x) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e0c5034915c6f3174eaa98e09a3e4654a3cf26c/pkg/roachpb/data.proto#L615)):*
<details><summary><i>Previously, lyang24 (Lanqing Yang) wrote…</i></summary><blockquote>

I like this approach. changing crbd_internal.go is not easy. However will this duplicate lock waiters display in crdb internal? i.e. we have two lock holders with shared strength both of them will have the same waiters it will resulted in duplicated waiter row in crdb internal?
</blockquote></details>

I think we might have to do this computation on the client, by changing `crdb_internal.go` after all. 
___
*[`pkg/roachpb/data.proto` line 615 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/108543#-NikUxYB8Nv8G285Ru-M:-NikUxYB8Nv8G285Ru-N:b-iuczub) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c34aded94759a8925e07c93623ca12ea85f1da3/pkg/roachpb/data.proto#L615)):*
> ```proto
>   repeated kv.kvserver.concurrency.lock.Waiter waiters = 6 [(gogoproto.nullable) = false];
>   // The mode the lock is being held at.
>   kv.kvserver.concurrency.lock.Mode lock_mode = 7;
> ```

Lock modes are an internal concept to the concurrency package which are primarily used for conflict resolution. We should return just the lock strength instead.
___
*[`pkg/sql/logictest/testdata/logic_test/select_for_share` line 56 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108543#-Ng7ksIf1_2Kjd01HYA1:-Nikbsxh6XAqAHDDE9JG:b-ovf07h) ([raw file](https://github.com/cockroachdb/cockroach/blob/d926f1e51e23d477bcf3027adc6beac11513e58a/pkg/sql/logictest/testdata/logic_test/select_for_share#L56)):*
<details><summary><i>Previously, lyang24 (Lanqing Yang) wrote…</i></summary><blockquote>

Hey @arulajmani, I have addressed the comments do you mind take another look? Also one of the concern of decoupling LockStateInfo to KeyLocks is the waiters of shared locks may become duplicated rows on crdb_internal.cluster_locks. I did do a hack to prevent it but now individual lockStateInfo of the same shared lock may not have the correct waiter information.
</blockquote></details>

The diff on this file looks good. Thanks for addressing the TODO! 

I'll leave a comment re: hack to prevent duplicate rows closer to that change.


<!-- Sent from Reviewable.io -->

lyang24 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108543)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`pkg/roachpb/data.proto` line 615 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/108543#-Ni6QphL793T0894dsFw:-Nj61d6_9Pzg-Co6JEUt:b-xu6rf8) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e0c5034915c6f3174eaa98e09a3e4654a3cf26c/pkg/roachpb/data.proto#L615)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I think we might have to do this computation on the client, by changing `crdb_internal.go` after all.
</blockquote></details>

yea client filtering is the way to go my previous logic had a flaw if I set the lock waiter of any shared lock holder to empty the lock will display as not contented even though there are waiters for the shared lock. I came up with a poor man's version of filtering on crdb internal, I think it will work because of the way we generated the lockStateInfo by spans. Please take another look and let me know.
___
*[`pkg/sql/logictest/testdata/logic_test/select_for_share` line 56 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108543#-Ng7ksIf1_2Kjd01HYA1:-Nj62kec9Wl9PxeykvZj:b8mgym8) ([raw file](https://github.com/cockroachdb/cockroach/blob/d926f1e51e23d477bcf3027adc6beac11513e58a/pkg/sql/logictest/testdata/logic_test/select_for_share#L56)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

The diff on this file looks good. Thanks for addressing the TODO! 

I'll leave a comment re: hack to prevent duplicate rows closer to that change.
</blockquote></details>

Thank you for the feedback, I have addressed the comments please take another look and let me know.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (approved):
The change looks good. Should be good to merge once the commit message has been updated. Thanks for working through various iterations on this @lyang24! 

> Reviewed 1 of 5 files at r4, 6 of 6 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108543#-:-Nj9TGupEQoN4U1n4KcH:b4mbng0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lyang24 and @nvanbenschoten)

___
*[`-- commits` line 9 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/108543#-Nj9FqgZ-4dzqCrwcbhW):*
nit: let's drop this line from the commit message. We generally don't modify the commit message from one review iteration to the next -- instead, we keep it as we expect it to look when the commit is actually merged.
___
*[`pkg/roachpb/data.proto` line 615 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/108543#-Ni6QphL793T0894dsFw:-Nj9T-Ov8sL9an15ZtgJ:b-9ycl86) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e0c5034915c6f3174eaa98e09a3e4654a3cf26c/pkg/roachpb/data.proto#L615)):*
<details><summary><i>Previously, lyang24 (Lanqing Yang) wrote…</i></summary><blockquote>

yea client filtering is the way to go my previous logic had a flaw if I set the lock waiter of any shared lock holder to empty the lock will display as not contented even though there are waiters for the shared lock. I came up with a poor man's version of filtering on crdb internal, I think it will work because of the way we generated the lockStateInfo by spans. Please take another look and let me know.
</blockquote></details>

Yup, the new logic works. Good catch on the `contended` bug.


<!-- Sent from Reviewable.io -->

lyang24 reviewed (commented):
Thank you fixed commit message

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108543#-:-NjA4Dd5EbhfFdwYS_7m:b-5lri2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)



<!-- Sent from Reviewable.io -->

lyang24 commented:
TFTR!

bors r+

arulajmani reviewed (approved):
I'll run it on your behalf. Thanks again!

bors r+

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108543#-:-NjDKLxY-3u5uBMCNBwL:blgengq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lyang24 and @nvanbenschoten)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
@arulajmani @nvanbenschoten do we want to backport this to 23.2?

yuzefovich commented:
I included this PR into the backport of #114617.
